gematria/granite/graph_builder_model_inference.cc

```diff
@@ -346,7 +424,48 @@ GraphBuilderModelInference::FromTfLiteModel(
   }
   llvm::Expected<std::unique_ptr<tflite::Interpreter>> interpreter =
       CreateInterpreter(*tflite_model);
-  if (auto error = interpreter.takeError()) return error;
+  if (llvm::Error error = interpreter.takeError()) return error;
+
+  // Get a mapping between the input tensors used in the model and their
+  // corresponding indices.
+  auto input_tensor_to_idx =
+      std::make_unique<std::vector<int>>(kNumPossibleInputTensors, -1);
+  for (int idx : (*interpreter)->inputs()) {
+    const std::string_view name = (*interpreter)->GetInputName(idx);
+    input_tensor_to_idx->at(static_cast<int>(GetInputTensorFromName(name))) =
+        idx;
+  }
+
+  // Checks whether the model has all input tensors required to have for all
+  // graph builder models.
+  std::vector<InputTensor> missing_inputs;
+  for (const InputTensor input : kRequiredInputTensors) {
+    if (input_tensor_to_idx->at(static_cast<int>(input)) == -1)
+      missing_inputs.push_back(input);
+  }
+  if (!missing_inputs.empty()) {
+    std::stringstream buffer;
+    buffer << "Model is missing input tensors. ";
+    for (InputTensor missing_input : missing_inputs) {
+      buffer << GetInputTensorName(missing_input) << ", ";
+    }
+    buffer.seekp(-2, std::ios_base::end);
+    buffer << " were expected but not found";
+    return llvm::createStringError(llvm::errc::invalid_argument, buffer.str());
+  }
+
+  // Ensures no unexpected input tensors are present.
+  const int num_input_tensors = (*interpreter)->inputs().size();
+  const bool uses_deltas = input_tensor_to_idx->at(static_cast<int>(
+                               InputTensor::kDeltaBlockIndexTensor)) != -1;
+  const int num_expected_input_tensors =
+      kNumRequiredInputTensors + int(uses_deltas);
+  if (num_input_tensors != num_expected_input_tensors) {
+    return llvm::createStringError(
+        llvm::errc::invalid_argument,
+        "Unexpected number of input tensors. Expected %d, found %d.",
+        num_expected_input_tensors, num_input_tensors);
+  }
```

Comment:
I've just added that as part of handling the case where the name passed to `GetInputTensorFromName` is not valid, so the input tensor count check is no longer strictly required. But I can leave it in, since it is cheap and seems like a nice exit point after doing all the checks on the model and reporting all of the extra and missing tensors.

Reply:
With returning an error for unexpected tensors right above, I think we can remove this part.

Or, maybe even nicer, we can collect missing input tensor names and unexpected input tensor names, and if any of these is non-empty, then return them both in a single error message here.

But if we do the checks above correctly, the checks for the number of tensors would not provide any additional safety.